### ISSUE TITLE
Dataframe assignment breaks Timestamp series (possibly due to rounding)

### ISSUE THREAD

**Issue by coells** [pandas-dev/pandas#29482]
#### Code Sample

```python
import pandas as pd

print(pd.__version__)

df = pd.DataFrame({
    't': [pd.Timestamp('2017-05-22 22:28:35.489689'),
          pd.Timestamp('2017-05-22 22:28:36.086933'),
          pd.Timestamp('2017-05-22 22:28:37.000001')]
})

print(df)
print()

mask = [True, True, True]

t = df['t'].values

print(t)
print()

df.loc[mask, 't'] = t

print(df)
print()
```

#### Problem description

Upgrade from Pandas 0.24.1 to 0.25.3 causes a bug in how a Timestamp series/ndarray is handled. The bug seems to be present at least since 0.25.1.

The code above takes a subset of dataframe column and stores it as separate series or ndarray. After the assignment back to the dataframe, nanosecond precision is broken as shows the script output.

```
0.25.3
                           t
0 2017-05-22 22:28:35.489689
1 2017-05-22 22:28:36.086933
2 2017-05-22 22:28:37.000001

['2017-05-22T22:28:35.489689000' '2017-05-22T22:28:36.086933000'
 '2017-05-22T22:28:37.000001000']

                              t
0 2017-05-22 22:28:35.489689088
1 2017-05-22 22:28:36.086932992
2 2017-05-22 22:28:37.000001024
```

#### Expected Output

The "made up" nanoseconds are not random, hence it seems to be some kind of rounding. I have tried to provide as-simple-as-possible case based on our code that combines further data handling. 
We do use nanoseconds, but sometimes Timestamp is rounded to microseconds and we expect Pandas to keep it so.

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.6.8.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.4.113-el6.x86_64.lime.1
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.14.1
pytz             : 2019.3
dateutil         : 2.8.0
pip              : 18.1
setuptools       : 39.2.0
Cython           : 0.29.6
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.2.1
html5lib         : 1.0.1
pymysql          : None
psycopg2         : None
jinja2           : 2.10
IPython          : 7.4.0
pandas_datareader: None
bs4              : 4.6.3
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : 4.2.1
matplotlib       : 2.2.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.1.0
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
</details>

**Comment by jmakovicka** [comment]
AFAICS the problem goes as follows:

- In Block.setitem, the datetime64[ns] value arg is first coerced into int64
- then, in block.py:855 self._can_hold_element(value) returns false
- int64 is promoted to float64, which corrupts the nanoseconds

The promotion in block.py:855 should be probably avoided if self.is_datetime == True.

**Comment by jmakovicka** [comment]
This seems fixed in Pandas 1.0

**Comment by coells** [comment]
Agreed, closing as the bug seems to be fixed in 1.0
